clusterloader2/pkg/measurement/common/ooms_tracker.go

```diff
@@ -0,0 +1,200 @@
+		klog.Fatalf("Cannot register %s: %v", clusterOomsTrackerName, err)
+	}
+}
+
+func createClusterOomsTrackerMeasurement() measurement.Measurement {
+	fieldSelector := fields.Set{
+		"reason": oomEventReason,
+	}.AsSelector().String()
+	return &clusterOomsTrackerMeasurement{
+		selector: &measurementutil.ObjectSelector{
+			FieldSelector: fieldSelector,
+			Namespace:     metav1.NamespaceAll,
+		},
+		msgRegex: regexp.MustCompile(oomEventMsgRegexPattern),
+	}
+}
+
+type clusterOomsTrackerMeasurement struct {
+	// lock for synchronized access to ooms field
+	lock      sync.Mutex
+	selector  *measurementutil.ObjectSelector
+	msgRegex  *regexp.Regexp
+	isRunning bool
+	stopCh    chan struct{}
+	ooms      []oomEvent
+}
+
+type oomEvent struct {
+	Node          string    `json:"node"`
+	Process       string    `json:"process"`
+	ProcessMemory string    `json:"memory"`
+	ProcessID     int       `json:"pid"`
+	Time          time.Time `json:"time"`
+}
+
+func (m *clusterOomsTrackerMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	clusterOomsTrackerEnabled, err := util.GetBoolOrDefault(config.Params, clusterOomsTrackerEnabledFlagName, false)
+	if err != nil {
+		return nil, err
+	}
+	if !clusterOomsTrackerEnabled {
+		klog.Info("skipping tracking of OOMs in the cluster")
+		return []measurement.Summary{}, nil
+	}
+
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		return []measurement.Summary{}, m.start(config.ClusterFramework.GetClientSets().GetClient())
+	case "gather":
+		summary, err := m.gather()
+		if err != nil {
+			return summary, err
+		}
+		return summary, m.validate()
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func (m *clusterOomsTrackerMeasurement) Dispose() {
+	m.stop()
+}
+
+func (m *clusterOomsTrackerMeasurement) String() string {
+	return clusterOomsTrackerName
+}
+
+func (m *clusterOomsTrackerMeasurement) start(c clientset.Interface) error {
+	if m.isRunning {
+		klog.Infof("%s: cluster OOMs tracking measurement already running", m)
+		return nil
+	}
+	klog.Infof("%s: starting cluster OOMs tracking measurement...", m)
+	m.isRunning = true
+	m.stopCh = make(chan struct{})
+	m.ooms = make([]oomEvent, 0)
+	// Watching for OOM events from node-problem-detector below.
+	i := informer.NewInformer(
+		c,
+		"events",
+		m.selector,
+		m.handleOomEvent,
+	)
+	return informer.StartAndSync(i, m.stopCh, informerTimeout)
+}
+
+func (m *clusterOomsTrackerMeasurement) stop() {
+	if m.isRunning {
+		m.isRunning = false
+		close(m.stopCh)
+	}
+}
+
+func (m *clusterOomsTrackerMeasurement) gather() ([]measurement.Summary, error) {
+	klog.Infof("%s: gathering cluster OOMs tracking measurement", clusterOomsTrackerName)
+	if !m.isRunning {
+		return nil, fmt.Errorf("measurement %s has not been started", clusterOomsTrackerName)
+	}
+
+	m.stop()
+
```

Comment:
I put it inside "gather" case of a switch statement inside `Execute` function instead to prevent a situation in which `m.gather()` reads a different `m.ooms` field than `m.validate()`. Done.